_posts/2020-08-28-porting-namecoin-in-tor-browser-to-windows.md

```diff
@@ -0,0 +1,108 @@
+---
+layout: post
+title: "Porting Namecoin in Tor Browser to Windows"
+author: Jeremy Rand
+tags: [News]
+---
+
+As you no doubt remember [from 36C3]({{ "/2020/01/11/36c3-summary.html" | relative_url }}), the GNU/Linux version of Tor Browser Nightly comes with Namecoin support included.  
+While we've received significant test feedback (overwhelmingly positive), it's been pointed out that supporting Windows would enable additional test feedback, since not everyone has a GNU/Linux machine to test things on.  
+So, I've been porting the Namecoin support in Tor Browser Nightly to Windows.
+
+## StemNS and Tor Bootstrap
+
+First off, since this endeavor was going to involve some changes to StemNS, I figured this was a good opportunity to investigate an odd bug that was happening in StemNS sometimes.  
+I had noticed last year that in a fresh Tor Browser install, if Namecoin was enabled on the first run, Tor Launcher would indicate that Tor bootstrap had stalled at the "loading authority certificates" stage.  
+I found that spamming the Tor Launcher buttons to cancel and retry connecting would usually make the connection succeed after 5-10 tries, which seemed to suggest a race condition.  
+After quite a lot of manual inspection of StemNS logs, I found something odd.  
+The Tor control spec states that the `__LeaveStreamsUnattached` config option will cause all streams to wait for the controller (StemNS in this case) to attach them.  
+However, empirically, this was not what was happening.  
+Streams created as a result of user traffic (e.g. Firefox or Electrum-NMC) were waiting for StemNS to attach them, but streams created by Tor's bootstrap were still automatically being attached?  The Tor control spec goes on to say the following:
+
+> Attempting to attach streams
+> via TC when "__LeaveStreamsUnattached" is false may cause a race between
+> Tor and the controller, as both attempt to attach streams to circuits.}
+> 
+> You can try to attachstream to a stream that
+> has already sent a connect or resolve request but hasn't succeeded
+> yet, in which case Tor will detach the stream from its current circuit
+> before proceeding with the new attach request.}
+
+Well that certainly explains what was happening.  
```

Comment:
I think the wording is somewhat unclear here. I had to read it a few times to understand it. It was hard to understand at first whether your description of the streams' behavior was the observed behavior or the intended behavior. I suggest you reword it to something like this:

> After quite a lot of manual inspection of StemNS logs, I found something odd: While the Tor control spec claims that the `__LeaveStreamsUnattached` config option will cause all streams to wait for the controller (StemNS in this case) to attach them, this was empirically not what was happening.
> Rather, streams created as a result of user traffic (e.g. Firefox or Electrum-NMC) were waiting for StemNS to attach them, but streams created by Tor's bootstrap were still automatically being attached.  The Tor control spec goes on to say the following:

(quote)

> This certainly explained what had been happening.